codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/GoMiddlewareGenerator.java

```diff
@@ -0,0 +1,201 @@
+                                .build())
+                        .build())
+                .outputType(SymbolReference.builder()
+                        .symbol(SymbolUtils.createValueSymbolBuilder("middleware.SerializeOutput")
+                                .addReference(SymbolUtils.createNamespaceReference(GoDependency.SMITHY_MIDDLEWARE))
+                                .build())
+                        .build())
+                .handlerType(SymbolReference.builder()
+                        .symbol(SymbolUtils.createValueSymbolBuilder("middleware.SerializeHandler")
+                                .addReference(SymbolUtils.createNamespaceReference(GoDependency.SMITHY_MIDDLEWARE))
+                                .build())
+                        .build())
+                .build();
+    }
+
+    public static GoMiddlewareGenerator newDeserializeMiddleware(String identifier) {
+        return builder()
+                .identifier(identifier)
+                .handleMethodName("HandleDeserialize")
+                .inputType(SymbolReference.builder()
+                        .symbol(SymbolUtils.createValueSymbolBuilder("middleware.DeserializeInput")
+                                .addReference(SymbolUtils.createNamespaceReference(GoDependency.SMITHY_MIDDLEWARE))
+                                .build())
+                        .build())
+                .outputType(SymbolReference.builder()
+                        .symbol(SymbolUtils.createValueSymbolBuilder("middleware.DeserializeOutput")
+                                .addReference(SymbolUtils.createNamespaceReference(GoDependency.SMITHY_MIDDLEWARE))
+                                .build())
+                        .build())
+                .handlerType(SymbolReference.builder()
+                        .symbol(SymbolUtils.createValueSymbolBuilder("middleware.DeserializeHandler")
+                                .addReference(SymbolUtils.createNamespaceReference(GoDependency.SMITHY_MIDDLEWARE))
+                                .build())
+                        .build())
+                .build();
+    }
+
+    public void writeMiddleware(GoWriter writer, BiConsumer<GoMiddlewareGenerator, GoWriter> handlerBodyConsumer) {
+        writeMiddleware(writer, (m, w) -> {
+        }, handlerBodyConsumer);
+    }
+
+    public void writeMiddleware(GoWriter writer, BiConsumer<GoMiddlewareGenerator, GoWriter> fieldConsumer,
+                                BiConsumer<GoMiddlewareGenerator, GoWriter> handlerBodyConsumer) {
+        writer.addUseImports(CONTEXT_TYPE);
+        writer.addUseImports(METADATA_TYPE);
+        writer.addUseImports(inputType);
+        writer.addUseImports(outputType);
+        writer.addUseImports(handlerType);
+
+        Symbol middlwareSym = SymbolUtils.createPointableSymbolBuilder(identifier).build();
+
+        writer.openBlock("type $L struct {", "}", middlwareSym, () -> {
```

Comment:
Each of these blocks should have a comment explaining what the block being created does.